plugin/trino-iceberg/src/test/java/io/trino/plugin/iceberg/BaseIcebergConnectorTest.java

```diff
@@ -6237,12 +6234,67 @@ protected OptionalInt maxTableNameLength()
         return OptionalInt.of(255 - 33);
     }
 
+    @Test
+    public void testSetPartitionedColumnType()
+    {
+        try (TestTable table = new TestTable(getQueryRunner()::execute, "test_set_partitioned_column_type_", "WITH (partitioning = ARRAY['part']) AS SELECT 1 AS id, CAST(123 AS integer) AS part")) {
+            assertUpdate("ALTER TABLE " + table.getName() + " ALTER COLUMN part SET DATA TYPE bigint");
+
+            assertThat(query("SELECT part FROM " + table.getName()))
+                    .matches("VALUES bigint '123'");
+            assertThat(query("SELECT id FROM " + table.getName() + " WHERE part = 123"))
+                    .isFullyPushedDown();
+            assertThat((String) computeScalar("SHOW CREATE TABLE " + table.getName()))
+                    .contains("partitioning = ARRAY['part']");
+        }
+    }
+
+    @Test
+    public void testSetTransformPartitionedColumnType()
+    {
+        try (TestTable table = new TestTable(getQueryRunner()::execute, "test_set_partitioned_column_type_", "WITH (partitioning = ARRAY['bucket(part, 10)']) AS SELECT CAST(123 AS integer) AS part")) {
+            assertUpdate("ALTER TABLE " + table.getName() + " ALTER COLUMN part SET DATA TYPE bigint");
+
+            assertThat(query("SELECT * FROM " + table.getName()))
+                    .matches("VALUES bigint '123'");
+            assertThat((String) computeScalar("SHOW CREATE TABLE " + table.getName()))
+                    .contains("partitioning = ARRAY['bucket(part, 10)']");
+        }
+    }
+
     @Override
     protected void verifyTableNameLengthFailurePermissible(Throwable e)
     {
         assertThat(e).hasMessageMatching("Failed to create file.*|Could not create new table directory");
     }
 
+    @Override
+    protected Optional<SetColumnTypeSetup> filterSetColumnTypesDataProvider(SetColumnTypeSetup setup)
+    {
+        if ((setup.sourceColumnType().equals("bigint") && setup.newColumnType().equals("integer")) ||
+                (setup.sourceColumnType().equals("decimal(5,3)") && setup.newColumnType().equals("decimal(5,2)")) ||
+                (setup.sourceColumnType().equals("varchar") && setup.newColumnType().equals("char(100)")) ||
+                (setup.sourceColumnType().equals("row(x int)") && setup.newColumnType().equals("row(y int)"))) {
```

Comment:
That's subjective, but i find the following somewhat easier to read

```java
switch ("%s -> %s".formatted(setup.sourceColumnType(), setup.newColumnType())) {
    case "bigint -> integer":
    case "decimal(5,3) -> decimal(5,2)":
    case "varchar -> char(100)":
    case "row(x int) -> row(y int)":
        // Iceberg allows updating column types if the update is safe. Safe updates are:
        // - int to bigint
        // - float to double
        // - decimal(P,S) to decimal(P2,S) when P2 > P (scale cannot change)
        // https://iceberg.apache.org/docs/latest/spark-ddl/#alter-table--alter-column
        return Optional.of(setup.asUnsupported());

    case "varchar(100) -> varchar(50)":
        // Iceberg connector ignores the varchar length
        return Optional.empty();
}
```

Reply:
of course, this will become eve nicer when we can use record patterns in a switch (https://openjdk.org/jeps/405)